src/mds/locks.c

```diff
@@ -3,9 +3,18 @@
 #include <string.h>
 #include <fcntl.h>
 
-#include "include/ceph_fs.h"
 #include "locks.h"
 
+/* Duplicated from ceph_fs.h, which we cannot include into a C file.  */
+#define CEPH_CAP_GSHARED     1  /* client can reads */
+#define CEPH_CAP_GEXCL       2  /* client can read and update */
+#define CEPH_CAP_GCACHE      4  /* (file) client can cache reads */
+#define CEPH_CAP_GRD         8  /* (file) client can read */
+#define CEPH_CAP_GWR        16  /* (file) client can write */
+#define CEPH_CAP_GBUFFER    32  /* (file) client can buffer writes */
+#define CEPH_CAP_GWREXTEND  64  /* (file) client can extend EOF */
+#define CEPH_CAP_GLAZYIO   128  /* (file) client can perform lazy io */
```

Comment:
@jtlayton should we add a backpointer in ceph_fs.h? or put these in a separate header?

Reply:
I think putting them in a separate header would be best.

Reply:
Since these defines are shared with kernel code, that would mean this separate header would also have to go into the kernel, right?

Another alternative might be to just rewrite this file in C++ -- I didn't attempt to do that since the file makes significant use of the C99 designated initializer feature that doesn't exist in C++, so would probably require a complete rewrite ...

Reply:
Ideally, yes, this would be a kernel-provided header, but we'll need to be able to cope with building with kernel-header packages that may not provide it for quite a while anyway.

I'd start by just moving this into a new header within the ceph tree for now, and then we once we have a kernel release that provides this header, we can add a cmake test and only use the in-tree header when the kernel doesn't provide it.

Reply:
Yes, we definitely don't want the kernel to provide anything to userspace.  It has always had an outdated copy of these "shared" headers, which was only partially synced up when needed.  The ceph repo has always been canonical.

I'm not sure we need another header just for CEPH_CAP_* constants.  It sounds like the subset that is intended to be moved to the new header is completely arbitrary -- "things that are needed to get that random .c file over there to compile"...

Duplicating 8 set-in-stone constants with an explanation in a comment doesn't seem too bad to me.  Barring that, I would vote for rewriting lock.c: wrap those arrays into classes and initialize them in constructors to work around the lack of designated initializers in C++.

Reply:
Agree... let's just add a corresponding comment in ceph_fs.h referencing the duplicated definitions here.

Rewriting locks.c would be a pain in the butt and probably not a great use of time.  The ore convenient C99 declaration style is why its C in the first place.

Reply:
Heh, looks like C++20 is bringing some limited support for designated initializers, but subscript designators is one of the things that will not be supported (because of conflicts with with ​lambda-expression​ syntax):

```
int arr[3] = {[1] = 5};        ​// valid C, invalid C++
```

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0329r4.pdf

Reply:
Added comment in ceph_fs.h as requested.